### PR TITLE
[SPARK-41043][SQL] Rename the error class `_LEGACY_ERROR_TEMP_2429` to `NUM_COLUMNS_MISMATCH`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to assign the proper name `NUM_COLUMNS_MISMATCH ` to the legacy error class `_LEGACY_ERROR_TEMP_2429 `, and modify test suite to use `checkError()` which checks the error class name, context and etc.

### Why are the changes needed?
Proper name improves user experience w/ Spark SQL.

### Does this PR introduce _any_ user-facing change?
Yes, the PR changes an user-facing error message. 

### How was this patch tested?
By running the modified tests:
```
$ build/sbt "test:testOnly *DataFrameSetOperationsSuite"
$ PYSPARK_PYTHON=python3 build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite"
```

### CODE REVIEW

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -94,6 +94,11 @@
     ],
     "sqlState" : "22005"
   },
+  "COLUMNS_NUM_MISMATCH" : {
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:97]: `NUM_COLUMNS_MISMATCH`

**Review by cloud-fan — APPROVED** [review]

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -94,6 +94,11 @@
...
+    "message" : [
+      "<operator> can only be performed on tables with the same number of columns, but the first table has <refColumnsNum> columns and the <invalidOrdinalNum> table has <invalidColumnsNum> columns."
```
Comment (srielau) [inline core/src/main/resources/error/error-classes.json:99]:
```suggestion
      "<operator> can only be performed on tables with the same number of columns, but the first table has <refNumColumns> columns and the <invalidOrdinalNum> table has <invalidNumColumns> columns."
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -541,12 +541,12 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
               // Check the number of columns
               if (child.output.length != ref.length) {
                 e.failAnalysis(
-                  errorClass = "_LEGACY_ERROR_TEMP_2429",
+                  errorClass = "COLUMNS_NUM_MISMATCH",
```
Comment (srielau) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:544]:
```suggestion
                  errorClass = "NUM_COLUMNS_MISMATCH",
```

File: sql/core/src/test/resources/sql-tests/results/except-all.sql.out [hunk 4]
```diff
@@ -228,12 +228,12 @@ struct<>
 -- !query output
 org.apache.spark.sql.AnalysisException
 {
-  "errorClass" : "_LEGACY_ERROR_TEMP_2429",
+  "errorClass" : "COLUMNS_NUM_MISMATCH",
```
Comment (srielau) [inline sql/core/src/test/resources/sql-tests/results/except-all.sql.out:231]:
```suggestion
  "errorClass" : "NUM_COLUMNS_MISMATCH",
```

File: sql/core/src/test/resources/sql-tests/results/except-all.sql.out [hunk 5]
```diff
@@ -228,12 +228,12 @@ struct<>
...
   "messageParameters" : {
-    "firstColNum" : "1",
-    "nColNum" : "2",
-    "nTab" : "second",
-    "operator" : "ExceptAll"
+    "invalidColumnsNum" : "2",
+    "invalidOrdinalNum" : "second",
+    "operator" : "EXCEPTALL",
+    "refColumnsNum" : "1"
```
Comment (srielau) [inline sql/core/src/test/resources/sql-tests/results/except-all.sql.out:236]:
```suggestion
    "invalidNumColumns" : "2",
    "invalidOrdinalNum" : "second",
    "operator" : "EXCEPTALL",
    "refNumColumns" : "1"
```

**Review by srielau — APPROVED** [review]
With comments

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:97]: Not sure about this. For me, it means columns that contains numbers but columns number. If `NUM` comes first, it should `NUM_OF_COLUMNS_MISMATCH`, IMHO.

Comment (srielau) [reply core/src/main/resources/error/error-classes.json:97]: Here is the pattern have used throughout my career: for (colNum = 0; colNum < numCols; colNum++).
I have no objection to NUM_OF_COLUMNS_MISMATCH or NUMBER_OF_COLUMNS_MISMATCH. The error-class name is not too long.

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:97]: ok. I will rename it.

**Comment by MaxGekk** [comment]
Merging to master. All GAs passed.
Thank you, @srielau @cloud-fan @LuciferYang for review.

**Pull request closed without merging**